include/pybind11/pybind11.h

```diff
@@ -1954,25 +1954,49 @@ inline std::pair<decltype(internals::registered_types_py)::iterator, bool> all_t
     return res;
 }
 
-template <typename Iterator, typename Sentinel, bool KeyIterator, return_value_policy Policy>
+template <typename Access, return_value_policy Policy, typename Iterator, typename Sentinel, typename ValueType, typename... Extra>
```

Comment:
The struct doesn't explicitly depend on typename Access -- it makes me wonder if we need a new iterator_state per Access and Policy type?

Seems like we are creating Python surrogates (class_<state>) a few lines later, and maybe we want to differentiate the types because of that?

Reply:
Exactly. Using `Access` replaces `KeyIterator`, which could only express two possible access methods.

Reply:
Thanks. If we missed Access here then we would be reusing a type with a wrong Accessor() object; a similar reason for ValueType. Do you see why we need Sentinel or Extra?

Here is a way to make this dependency explicit, but it feels a bit too verbose:

```
template <typename Access, typename ValueType, ....>
struct iterator_state {
Access::result_type get() {
  return Access()(it);
}
typedef ValueType value_type;

};

class_(state).def(state & s) -> state::value_type {
...
return s.get();
}
```

Reply:
> Do you see why we need Sentinel or Extra?

Extra is for the same reason as Access: if you don't have it there, then two different calls to `make_iterator_impl` with different `Extra` won't work, because they'll share the same Python class and it'll only be defined correctly for the Extra given in the first call. iterator_state has a member of type Sentinel.

> Here is a way to make this dependency explicit, but it feels a bit too verbose:

I don't have any strong objections, but I also feel it doesn't really add anything (and is verbose). If you want me to implement that let me know.

Reply:
Good idea. Done.